README.md
# Dionysus

[![Discord](https://img.shields.io/discord/819436478865997874?label=discord)](https://discord.gg/dV7rp5F4uA)
[![GitHub all releases](https://img.shields.io/github/downloads/nopjmp/Dionysus/total)](https://github.com/nopjmp/Dionysus/releases)
[![GitHub contributors](https://img.shields.io/github/contributors/nopjmp/dionysus)](https://github.com/nopjmp/Dionysus/graphs/contributors) 
[![GitHub issues](https://img.shields.io/github/issues/nopjmp/Dionysus)](https://github.com/nopjmp/Dionysus/issues)

###### Maintained by contributors. Owner Discord: nopjmp#1337 

----

#### 1.12.2 [Paper](https://github.com/PaperMC/Paper) fork aimed at improving server performance for anarchy servers. [Latest Dev Build*](https://nightly.link/nopjmp/Dionysus/workflows/dev/dev/Dionysus-JDK17.zip)
###### Note: This fork is based off [Aikar's EMC Framework](https://github.com/starlis/empirecraft)

----

#### Used and trusted by:
- [2b2t.au](https://2b2t.au)
- [6b6t.org](https://6b6t.org)
- [oldfrog.org](https://oldfrog.org)

## Contributing

*Anyone can contribute. Just follow the below steps!* 

1. [Fork Dionysus](https://github.com/nopjmp/Dionysus/fork)
2. [Build Dionysus](https://github.com/nopjmp/Dionysus#building)
3. [Create Pull Request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork)

Assistant will be given to get pull requests working correctly.

## Server Admins

Dionysus uses the same paperclip jar system that Paper uses.  
You can download the latest release of Dionysus [here](https://github.com/nopjmp/Dionysus/releases/latest).   
You can also [build it yourself](https://github.com/nopjmp/Dionysus#building)

## Plugin Developers
###### *With each Dionysus update you must update your dependency.*

In order to use Dionysus as a dependency you must [build it yourself](https://github.com/nopjmp/Dionysus#building)      
This will add it to your local maven repository folder. Then add the following to your `pom.xml`:

#### Dionysus-API Maven Dependency:
```xml
<dependency>
    <groupId>dev.pomf.dionysus</groupId>
    <artifactId>dionysus-api</artifactId>
    <version>1.12.2-R0.1-SNAPSHOT</version>
    <scope>provided</scope>
</dependency>
```
#### Dionysus-Server Maven Dependency:
```xml
<dependency>
    <groupId>dev.pomf.dionysus</groupId>
    <artifactId>dionysus</artifactId>
    <version>1.12.2-R0.1-SNAPSHOT</version>
    <scope>provided</scope>
</dependency>
```

#### Local Maven Repository:

Windows: `C:\Users\<UserName>\.m2`
Linux: `/home/<UserName>/.m2`
Mac: `/Users/<UserName>/.m2`

```xml
<repository>
    <id>project.local</id>
    <name>project</name>
    <url>file:/Users/User/.m2/repository/</url>
</repository>
```

## Building

**Requirements:**

- You need `git` installed, with a configured user name and email.
  On windows you need to run from git bash.
- You need `maven` installed
- You need `jdk 1.8` to decompile and `jdk` 17+ installed to compile (and `jre` 17+ to run)
- Anything else that `paper` requires to build

**If all you want is a paperclip server jar run `./dionysus jar`**

#### Setting up `Dionysus-API` and `Dionysus-Server`

1. Run `./dionysus patch` in your project root.
2. Run `./dionysus build` to build the respective api and server jars.

#### Creating a patch

Patches are effectively just commits in either `Dionysus-API` or `Dionysus-Server`.

1. Create commit in `Dionysus-API` or `Dionysus-Server`.
2. Run `./dionysus rb` in your project root.`
3. Commit patch to `Dionysus` repo.

*Modifying commits will also modify its corresponding patch file.*

### License

The PATCHES-LICENSE file describes the license for api & server patches,
found in `./patches` and its subdirectories except when noted otherwise.

Everything else is licensed under the MIT license, except when note otherwise.
See [empirecraft](https://github.com/starlis/empirecraft) and [byof](https://github.com/electronicboy/byof)
for the license of material used/modified by this project.


CHANGELOG.md
<!-- SPDX-License-Identifier: MIT -->
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.1.16] - 2023-03-25

* remove two patches around hoppers
* fix missing code in lag compensated eating patch
* Backport ArrowDMG fix
* Add possible fix for TP exploit
* override paper 1.12.2's download mechanism
* Remove unused variable now....
* Switch from ConcurrentQueue to synchronized list for plugin compat
* Trim map after clear test...
* Fix Patch
* PlayerItemHeldEvent Add Slot
* Clean up Dionysus Config
* remove unnecessary lines
* rename patch
* Clean up linear region file patch
* start on linear files
* Update README.md
* Prevent IndexOutOfBound in window packet
* Add few helpful patches
* Prevent recursive chunk loading for join chunks
* Improve do not load chunks for light checks patch
* Few improvements
* Update java information
* Set chunk coord int pair patch author to original author
* Merge branch 'dev' of https://github.com/nopjmp/Dionysus into dev
* Improve fire spread patch
* Drop revert xray (#49)
* Add Alternate Current redstone implementation (#48)
* Further improve light queue and improve other methods as well
* Temporarily drop map chunk load patch
* Prevent vehicle move while awaiting teleport confirm
* Tuinity Optimise ActivationRange#activateChunkEntities
* Fixup stats system with new long change
* Fix backwards compat with long change
* Migrate stats to long and improve stats map performance
* Migrate stats to long and improve stats map performance
* Use fastutil for entity tracker
* Add faster NBT compare options and improve guessSize patch
* Use FastThreadLocalThread for netty IO threads (#43)
* Prepare guessSize patch for future code for patches
* Add crash comment to config
* Fix a bug in disabling mojang server communication (#42)
* Add very dirty async world tick patch
* Fixup previous patch
* Backport Purpur's Alternative Keepalive Handling (#41)
* Add option to disable communication with mojang servers in offline mode (#40)
* Tiny cleanup of donkey patch
* Add option to disable duplicate log
* Fix banner color changing
* [ci skip] add nightly link
* Improve position iterator performance
* Port some other patches and improve performance (#39)
* Improve NeonPaper chest patch
* Improve fire spread checks
* Add slime setSize patch
* Improve guesssize method
* Optimize passenger list with fastutil
* Guava to caffeine
* Add dependency on caffeine
* Add a few extra patches
* Remove not needed tab change for update dependencies
* Fix-MC-161754
* Remove CHM patch
* Sort patches differently and delete comments from one patch
* Backport newer config system
* Add client sign limit
* Add jitpack way earlier
* Fix zstd dependency
* Add annotations to dependency commit
* Delete 0125-Dont-Tick-Dead-Players.patch
* Port even more patches
* Port few extra patches from FlamePaper, TacoSpigot and Reaper
* Port first few patches from NeonPaper
* Drop flare patch in API as well
* Drop flare profiler
* Paper Backport: Lag compensate eating
* Update sentry and allow concurrent book meta access
* Add a bit better handling for ZSTD config
* Fixup bukkit API in cases where data would still be written to the internal object
* Update Sentry and add pufferfish credits (#31)
* Port patches from Airplane and Pufferfish (#25)

## [0.1.15] - 2022-06-18

### Changes
* Improve merge checks
* Fix book meta speed for enchantments
* Fix build
* Add option to disable chunk loading past the world border
* Disable JNDI lookups in the default log config
* Improve performance of book meta pages

By only converting the pages when we need them, we save a lot of computing power for 99% of cases where pages are not needed and people just access the methods of ItemMeta.

* Improve performance of patch in more cases
* Fix small typos
* Fix error

## [0.1.14] - 2022-02-12

### Fixed
* Performance problem with the NBT size check

## [0.1.13] - 2022-02-05

### Added
* Add optimize armour stands from RemainingToast

## [0.1.12] - 2022-01-28

### Fixed
* CVE-2021-44832 - Apache Log4j2 updated to 2.17.1 (only affects JDBC Appender users)
* AntiBook Crash prevention caused issue with cloning books

## [0.1.11] - 2021-12-20

### Fixed
* CVE-2021-45105 - Apache Log4j2 updated to 2.17.0

### Removed
* Trove4J - outdated library that is slower than fastutils

## [0.1.10] - 2021-12-15

### Fixed

* Typos
* Don't count some map decorations
* TileEntity config didn't completely work

## [0.1.9] - 2021-12-14

### Fixed

* CVE-2021-45046 - Apache Log4j2 updated to 2.16.0

## [0.1.8] - 2021-12-12

### Fixed

* Log4j exploit by updating to 2.15.0
* O(N^2) Advancement Daya Player Interaction to be O(N)

### Added 
* configurable chunk dupe

## [0.1.7] - 2021-03-13

Hotfix for Structure generation

### Fixed

* Removed Xorshift on World.random
  * This is tied to structure generation for some reason in 1.12.2.
  * Will add a shared global random and do testing on moving entity and block behavior to a faster random source.

## [0.1.6] - 2021-03-12

### Added

* Add the ability to redact player ip addresses and coords
  * Banning a player by ip will still work, just not display in the logs.
  * Further customization should come later

### Fixed

* Fixed players getting kicked due to improper change for World.random
  * This was only in the dev builds but still was given out to some users

### Removed

* Networking System based on Velocity and Yatopia
  * This unfortunately touched a few areas that a version agnostic packet manipulation framework was using.
* Shapeless crafting from Airplane backport
  * This was only in dev builds and was a bad, broken port

## [0.1.5] - 2021-03-07

### Added

* Fair natural spawning
  * Searches around the players to avoid their locations for spawning in monsters if they are hitting the limit.
* Load Chunks for login asyncronously
* Removed EnumSet allocations in crafting checks
* Lighting system optimizations

scripts/init.sh
#!/usr/bin/env bash
# BEGIN config
FORK_NAME="Dionysus"
API_REPO=""
SERVER_REPO=""
PAPER_API_REPO=""
PAPER_SERVER_REPO=""
MCDEV_REPO=""
# END config

sourceBase=$(dirname $SOURCE)/../
cd "${basedir:-$sourceBase}"

basedir=$(pwd -P)
cd -


function bashColor {
if [ $2 ]; then
    echo -e "\e[$1;$2m"
else
    echo -e "\e[$1m"
fi
}
function bashColorReset {
    echo -e "\e[m"
}

function cleanupPatches {
    cd "$1"
    for patch in *.patch; do
        gitver=$(tail -n 2 $patch | grep -ve "^$" | tail -n 1)
        diffs=$(git diff --staged $patch | grep -E "^(\+|\-)" | grep -Ev "(From [a-z0-9]{32,}|\-\-\- a|\+\+\+ b|.index|Date\: )")

        testver=$(echo "$diffs" | tail -n 2 | grep -ve "^$" | tail -n 1 | grep "$gitver")
        if [ "x$testver" != "x" ]; then
            diffs=$(echo "$diffs" | tail -n +3)
        fi

        if [ "x$diffs" == "x" ] ; then
            git reset HEAD $patch >/dev/null
            git checkout -- $patch >/dev/null
        fi
    done
}
function pushRepo {
    if [ "$(git config minecraft.push-${FORK_NAME})" == "1" ]; then
    echo "Pushing - $1 ($3) to $2"
    (
        cd "$1"
        git remote rm emc-push > /dev/null 2>&1
        git remote add emc-push $2 >/dev/null 2>&1
        git push emc-push $3 -f
    )
    fi
}
function basedir {
    cd "$basedir"
}
function gethead {
    (
        cd "$1"
        git log -1 --oneline
    )
}


scripts/apply.sh
#!/usr/bin/env bash
# get base dir regardless of execution location
# resolve shell-specifics
case "$(echo "$SHELL" | sed -E 's|/usr(/local)?||g')" in
    "/bin/zsh")
        RCPATH="$HOME/.zshrc"
        SOURCE="${BASH_SOURCE[0]:-${(%):-%N}}"
    ;;
    *)
        RCPATH="$HOME/.bashrc"
        if [[ -f "$HOME/.bash_aliases" ]]; then
            RCPATH="$HOME/.bash_aliases"
        fi
        SOURCE="${BASH_SOURCE[0]}"
    ;;
esac

# get base dir regardless of execution location

while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
    DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
    SOURCE="$(readlink "$SOURCE")"
    [[ "$SOURCE" != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
done
. $(dirname $SOURCE)/init.sh
PS1="$"

paperVer=$(cat current-paper)
gpgsign="$(git config commit.gpgsign || echo "false")"
echo "Rebuilding Forked projects.... "
function applyPatch {
    what=$1
    what_name=$(basename $what)
    target=$2
    branch=$3
    patch_folder=$4

    cd "$basedir/$what"
    git fetch --all
    git branch -f upstream "$branch" >/dev/null

    cd "$basedir"
    if [ ! -d  "$basedir/$target" ]; then
        mkdir "$basedir/$target"
        cd "$basedir/$target"
        git init
	git config commit.gpgsign false
        git remote add origin "$5"
        cd "$basedir"
    fi
    cd "$basedir/$target"

    # Disable GPG signing before AM, slows things down and doesn't play nicely.
    # There is also zero rational or logical reason to do so for these sub-repo AMs.
    # Calm down kids, it's re-enabled (if needed) immediately after, pass or fail.
    git config commit.gpgsign false

    echo "Resetting $target to $what_name..."
    git remote rm upstream > /dev/null 2>&1
    git remote add upstream "$basedir/$what" >/dev/null 2>&1
    git checkout master 2>/dev/null || git checkout -b master
    git fetch upstream >/dev/null 2>&1
    git reset --hard upstream/upstream
    echo "  Applying patches to $target..."
    git am --abort >/dev/null 2>&1
    git am --3way --ignore-whitespace "$basedir/patches/$patch_folder/"*.patch
    if [ "$?" != "0" ]; then
        echo "  Something did not apply cleanly to $target."
        echo "  Please review above details and finish the apply then"
        echo "  save the changes with rebuildPatches.sh"
        exit 1
    else
        echo "  Patches applied cleanly to $target"
    fi
}

function enableCommitSigningIfNeeded {
    if [[ "$gpgsign" == "true" ]]; then
        git config commit.gpgsign true
    fi
}

    echo "Importing MC-DEV"
    ./scripts/importmcdev.sh "$basedir" || exit 1
(
    (applyPatch Paper/Paper-API ${FORK_NAME}-API HEAD api $API_REPO &&
    applyPatch Paper/Paper-Server ${FORK_NAME}-Server HEAD server $SERVER_REPO) || exit 1
    enableCommitSigningIfNeeded
) || (
    echo "Failed to apply patches"
    enableCommitSigningIfNeeded
    exit 1
) || exit 1


scripts/upstream.sh
#!/usr/bin/env bash
# get base dir regardless of execution location
SOURCE="${BASH_SOURCE[0]}"
while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
    DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
    SOURCE="$(readlink "$SOURCE")"
    [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
done
. $(dirname $SOURCE)/init.sh

git submodule update --init --recursive

if [[ "$1" == up* ]]; then
    (
        cd "$basedir/Paper/"
		git fetch && git reset --hard origin/ver/1.12.2
        cd ../
        git add Paper
    )
fi

paperVer=$(gethead Paper)

minecraftserverurl=$(cat "$basedir"/Paper/work/BuildData/info.json | grep serverUrl | cut -d '"' -f 4)
minecraftversion=$(cat "$basedir"/Paper/work/BuildData/info.json | grep minecraftVersion | cut -d '"' -f 4)
decompiledir="$basedir/Paper/work/Minecraft/$minecraftversion"
jarpath="$decompiledir/$minecraftversion"

if [ ! -f  "$jarpath.jar" ]; then
    echo "Downloading unmapped vanilla jar..."
    mkdir -p "$decompiledir"
    curl -s -o "$jarpath.jar" "$minecraftserverurl"
    if [ "$?" != "0" ]; then
        echo "Failed to download the vanilla server jar. Check connectivity or try again later."
        exit 1
    fi
fi

cd "$basedir/Paper/"

./paper patch

cd "Paper-Server"
mcVer=$(mvn -o org.apache.maven.plugins:maven-help-plugin:2.1.1:evaluate -Dexpression=minecraft_version | sed -n -e '/^\[.*\]/ !{ /^[0-9]/ { p; q } }')

basedir
. $basedir/scripts/importmcdev.sh

version=$(echo -e "Paper: $paperVer\nmc-dev:$importedmcdev")
tag="${minecraftversion}-${mcVer}-$(echo -e $version | shasum | awk '{print $1}')"
echo "$tag" > "$basedir"/current-paper

"$basedir"/scripts/generatesources.sh

cd Paper/

function tag {
(
    cd $1
    if [ "$2" == "1" ]; then
        git tag -d "$tag" 2>/dev/null
    fi
    echo -e "$(date)\n\n$version" | git tag -a "$tag" -F - 2>/dev/null
)
}
echo "Tagging as $tag"
echo -e "$version"

forcetag=0
if [ "$(cat "$basedir"/current-paper)" != "$tag" ]; then
    forcetag=1
fi

tag Paper-API $forcetag
tag Paper-Server $forcetag

pushRepo Paper-API $PAPER_API_REPO $tag
pushRepo Paper-Server $PAPER_SERVER_REPO $tag



scripts/paperclip.sh
#!/usr/bin/env bash

# Copied from https://github.com/PaperMC/Paper/blob/d54ce6c17fb7a35238d6b9f734d30a4289886773/scripts/paperclip.sh
# License from Paper applies to this file

(
set -e
basedir="$(cd "$1" && pwd -P)"
workdir="$basedir/Paper/work"
localworkdir="$basedir/work"
mcver=$(cat "$workdir/BuildData/info.json" | grep minecraftVersion | cut -d '"' -f 4)
paperjar="$basedir/Dionysus-Server/target/dionysus-$mcver.jar"
vanillajar="$workdir/Minecraft/$mcver/$mcver.jar"

(
    cd "$localworkdir/Paperclip"
    mvn clean package "-Dmcver=$mcver" "-Dpaperjar=$paperjar" "-Dvanillajar=$vanillajar"
)
cp "$localworkdir/Paperclip/assembly/target/paperclip-${mcver}.jar" "$basedir/dionysus-paperclip.jar"

echo ""
echo ""
echo ""
echo "Build success!"
echo "Copied final jar to $(cd "$basedir" && pwd -P)/dionysus-paperclip.jar"
) || exit 1


scripts/importmcdev.sh
#!/usr/bin/env bash

SOURCE="${BASH_SOURCE[0]}"
while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
    DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
    SOURCE="$(readlink "$SOURCE")"
    [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
done
. $(dirname $SOURCE)/init.sh

workdir="$basedir"/Paper/work
minecraftversion=$(cat "$basedir"/Paper/work/BuildData/info.json | grep minecraftVersion | cut -d '"' -f 4)
decompiledir=$workdir/Minecraft/$minecraftversion

nms="net/minecraft/server"
export MODLOG=""
cd "$basedir"

function containsElement {
    local e
    for e in "${@:2}"; do
        [[ "$e" == "$1" ]] && return 0;
    done
    return 1
}

export importedmcdev=""
function import {
    if [ -f "$basedir/Paper/Paper-Server/src/main/java/net/minecraft/server/$1.java" ]; then
        echo "ALREADY IMPORTED $1"
        return 0
    fi
    export importedmcdev="$importedmcdev $1"
    file="${1}.java"
    target="$basedir/Paper/Paper-Server/src/main/java/$nms/$file"
    base="$decompiledir/$nms/$file"

    if [[ ! -f "$target" ]]; then
        export MODLOG="$MODLOG  Imported $file from mc-dev\n";
        echo "$(bashColor 1 32) Copying $(bashColor 1 34)$base $(bashColor 1 32)to$(bashColor 1 34) $target $(bashColorReset)"
        cp "$base" "$target"
    else
        echo "$(bashColor 1 33) UN-NEEDED IMPORT STATEMENT:$(bashColor 1 34) $file $(bashColorReset)"
    fi
}


(
    cd Paper/Paper-Server/
    lastlog=$(git log -1 --oneline)
    if [[ "$lastlog" = *"Tuinity-Extra mc-dev Imports"* ]]; then
        git reset --hard HEAD^
    fi
)


files=$(cat patches/server/* | grep "+++ b/src/main/java/net/minecraft/server/" | sort | uniq | sed 's/\+\+\+ b\/src\/main\/java\/net\/minecraft\/server\///g' | sed 's/.java//g')

nonnms=$(cat patches/server/* | grep "create mode " | grep -Po "src/main/java/net/minecraft/server/(.*?).java" | sort | uniq | sed 's/src\/main\/java\/net\/minecraft\/server\///g' | sed 's/.java//g')

for f in $files; do
    containsElement "$f" ${nonnms[@]}
    if [ "$?" == "1" ]; then
        if [ ! -f "$basedir/Paper/Paper-Server/src/main/java/net/minecraft/server/$f.java" ]; then
            if [ ! -f "$decompiledir/$nms/$f.java" ]; then
                if [[ ! $f == Dionysus* ]]; then
                    echo "$(bashColor 1 31) ERROR!!! Missing NMS$(bashColor 1 34) $f $(bashColorReset)";
                fi
            else
                import $f
            fi
        fi
    fi
done

# Dionysus Modifications
import MathHelper
import EntityBat
import EnumDirection
import PacketCompressor
import PacketDecompressor
import PacketEncrypter
import PacketDecrypter
import StatisticWrapper
import BlockSign

########################################################
########################################################
########################################################
#              LIBRARY IMPORTS
# These must always be mapped manually, no automatic stuff
#
#             # group    # lib          # prefix               # many files

#importLibrary com.mojang datafixerupper com/mojang/datafixers/util Either.java
################
(
    cd Paper/Paper-Server/
    rm -rf nms-patches
    git add src -A
    echo -e "Tuinity-Extra mc-dev Imports\n\n$MODLOG" | git commit src -F -
    exit 0
)
